Add Reauth flow to Wallbox integration

hesselonline opened the pull request:
## Proposed change
I'm adding a reauthentication flow to the wallbox component. It's triggered by an HTML 403 error (so that will usually be a password change). The flow is triggered by calling ConfigEntryAuthFailed from the authentication method of the Wallbox class.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X ] The code change is tested and works locally.
- [X ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X ] There is no commented out code in this PR.
- [X ] I have followed the [development checklist][dev-checklist]
- [X ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

hesselonline committed:
Add Reauth flow to Wallbox integration

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox_add_reauth_flow

hesselonline committed:
Review comments processed

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox_add_reauth_flow

hesselonline committed:
Fixed tests

hesselonline committed:
Added test for reauth invalid

hesselonline commented:
@bdraco I added the discussed reauth flow to the wallbox component in this PR. Maybe you have time for a quick review? Thnx

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -122,20 +122,33 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
-    await wallbox_coordinator.async_config_entry_first_refresh()
+    except InvalidAuth:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_REAUTH, "entry_id": entry.entry_id},
+                data=entry.data,
+            )
+        )
+        return False
```
```suggestion
        raise ConfigEntryAuthFailed from ex
```

hesselonline replied:
Changed this in a new commit

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -122,20 +122,33 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        await wallbox_coordinator.async_validate_input()
 
-    await wallbox_coordinator.async_config_entry_first_refresh()
+    except InvalidAuth:
```
```suggestion
    except InvalidAuth as ex:
```

hesselonline replied:
Changed this in a new commit

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -122,20 +122,33 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        return False
 
-    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}})
-    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = wallbox_coordinator
```
Since you are only storing the coordinator, drop `CONF_CONNECTIONS` and store it as 

`hass.data[DOMAIN][entry.entry_id] = wallbox_coordinator`

hesselonline replied:
I changed this in a new commit, did however need the `  hass.data.setdefault(DOMAIN, {entry.entry_id: {}})` part, otherwise it would result in an error

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -122,20 +122,33 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}})
+        hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = wallbox_coordinator
+
+        for platform in PLATFORMS:
+
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
```
```suggestion
        hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -122,20 +122,33 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
-    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}})
-    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = wallbox_coordinator
+    else:
```
Drop the `else` and outdent below

hesselonline replied:
done

bdraco commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -47,13 +70,24 @@ async def async_step_user(self, user_input=None):
         try:
-            info = await validate_input(self.hass, user_input)
+            await self.async_set_unique_id(user_input["station"])
+            if not self._reauth_entry:
+                self._abort_if_unique_id_configured()
+                info = await validate_input(self.hass, user_input)
+
+            if not self._reauth_entry:
+                return self.async_create_entry(title=info["title"], data=user_input)
```
Combine this

hesselonline replied:
done

bdraco commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -47,13 +70,24 @@ async def async_step_user(self, user_input=None):
+
+            if not self._reauth_entry:
+                return self.async_create_entry(title=info["title"], data=user_input)
+            self.hass.config_entries.async_update_entry(
```
Abort if the station being reauthed is not the same account that was previous authorized. Reauth should never change the underlying config entry unique id

hesselonline replied:
I did this by checking `if user_input["station"] == self._reauth_entry.data[CONF_STATION]:` and doing a `errors["base"] = "reauth_invalid"` as an else. I don't know if there is a better / standard approach to this.

hesselonline committed:
Commit to compensate for timedrift, show changes
Compensating for timedrift in my devcontainer,
making a new commit with the right date/time. Requested changes
were done in a previous commit.

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox_add_reauth_flow

bdraco commented on homeassistant/components/wallbox/strings.json:
```diff
@@ -7,15 +7,23 @@
       "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
-      "unknown": "[%key:common::config_flow::error::unknown%]"
+      "unknown": "[%key:common::config_flow::error::unknown%]",
+      "reauth_invalid": "Re-authentication failed, Serial Number was invalid"
```
```suggestion
      "reauth_invalid": "Re-authentication failed, Serial Number does not match original"
```

bdraco replied:
That may still not be the most helpful message. Ideally we tell them what to do

hesselonline replied:
I'll use the text in your suggestion for now, might change it in the future if I find better words.

hesselonline replied:
I included this in the commit that fixes the merge conflicts

bdraco commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -162,3 +163,83 @@ async def test_form_validate_input(hass):
+    await setup_integration(hass)
+    assert entry.state == config_entries.ConfigEntryState.LOADED
+
+    with requests_mock.Mocker() as mock_request:
+        mock_request.get(
+            "https://api.wall-box.com/auth/token/user",
+            text='{"jwt":"fakekeyhere","user_id":12345,"ttl":145656758,"error":false,"status":200}',
+            status_code=200,
+        )
+        mock_request.get(
+            "https://api.wall-box.com/chargers/status/12345",
+            json=test_response,
+            status_code=200,
+        )
+
```
For a future PR:

Generally we prefer to mock the library calls themselves and test the api in the library.

This avoids having to update both sets of tests (the lib and Home Assistant) when the endpoint or API changes.

hesselonline replied:
OK, I will look into this

bdraco reviewed (commented):
Looks good.  One small comment about how to tell the user the reauth doesn't match

bdraco reviewed (changes requested):
Please resolve conflicts.  Thanks 👍

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox_add_reauth_flow
also change string for invalid serial

bdraco commented:
Test failure is unrelated.  I'll restart the CI

bdraco commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -36,6 +36,29 @@ async def validate_input(hass: core.HomeAssistant, data):
+            self.context["entry_id"]
+        )
+
+        self._reauth_schema = {
+            vol.Required(
+                CONF_STATION, default=self._reauth_entry.data[CONF_STATION]
+            ): str,
+            vol.Required(
+                CONF_USERNAME, default=self._reauth_entry.data[CONF_USERNAME]
+            ): str,
+            vol.Required(CONF_PASSWORD): str,
+        }
```
I don't see where this is used

hesselonline replied:
I agree, that can go, I removed it.

hesselonline committed:
remove reauth schema

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -112,18 +106,18 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    await wallbox_coordinator.async_config_entry_first_refresh()
 
-    for platform in PLATFORMS:
+    hass.data.setdefault(DOMAIN, {entry.entry_id: {}})
+    hass.data[DOMAIN][entry.entry_id] = wallbox_coordinator
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = wallbox_coordinator
```

bdraco reviewed (commented):
One small tweak, and this looks good to merge 👍

hesselonline committed:
Update homeassistant/components/wallbox/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

MartinHjelmare commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -47,14 +59,27 @@ async def async_step_user(self, user_input=None):
 
         try:
-            info = await validate_input(self.hass, user_input)
+            await self.async_set_unique_id(user_input["station"])
```
Please only wrap the lines that can raise in the try... except block.

MartinHjelmare commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -47,14 +59,27 @@ async def async_step_user(self, user_input=None):
         return self.async_show_form(
-            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            step_id="user",
+            data_schema=STEP_USER_DATA_SCHEMA,
```
Normally we should only ask the user for the required info to re-authenticate in the reauth flow. We shouldn't ask the user again for info that should stay the same like station id.

MartinHjelmare commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -47,14 +59,27 @@ async def async_step_user(self, user_input=None):
+                info = await validate_input(self.hass, user_input)
+                return self.async_create_entry(title=info["title"], data=user_input)
+            if user_input["station"] == self._reauth_entry.data[CONF_STATION]:
+                self.hass.config_entries.async_update_entry(
```
We  should validate the new credentials in the reauth flow before updating the config entry data.

MartinHjelmare commented on homeassistant/components/wallbox/strings.json:
```diff
@@ -7,15 +7,23 @@
           "password": "[%key:common::config_flow::data::password%]"
         }
+      },
+      "reauth_confirm": {
```
We're missing this step.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
